lib/url-pattern/url-match-pattern-list.js

```diff
@@ -0,0 +1,88 @@
+var _ = require('../util').lodash,
+    PropertyList = require('../collection/property-list').PropertyList,
+    Url = require('../collection/url').Url,
+    UrlMatchPattern = require('./url-match-pattern').UrlMatchPattern,
+
+    MATCH_ALL_URLS = UrlMatchPattern.MATCH_ALL_URLS,
+
+    UrlMatchPatternList;
+
+_.inherit((
+  UrlMatchPatternList = function (parent, list) {
+      UrlMatchPatternList.super_.call(this, UrlMatchPattern, parent, list);
+  }), PropertyList);
+
+_.assign(UrlMatchPatternList.prototype, /** @lends UrlMatchPatternList.prototype */ {
+    toJSON: function() {
+        if (!this.count()) {
+            return [];
+        }
+
+        return _.map(this.members, function (member) {
+            return member.toJSON();
+        });
+    },
+
+    /**
+      * Tests the url string with the match pattern list provided to see if it matches any of it.
+      * Follows the https://developer.chrome.com/extensions/match_patterns pattern for pattern validation and matching
+      *
+      * @param {String=} [urlStr] The url string for which the proxy match needs to be done.
+      * @returns {Boolean=}
+      */
+    test: function(urlStr) {
+        /*
+         * Similar to the UrlMatchPattern.test, however instead of testing
+         * MATCH_ALL_URLS and Regex conditions serially with each of the pattern,
+         * this method first searches for MATCH_ALL_URLS in all patterns
+         * and then moves on to the slower Regex based searches.
+         */
+        var url,
+            matchAllUrlsPattern,
+            matchedSpecificPattern;
+
+        matchAllUrlsPattern = this.find(function(urlMatchPattern) {
+            return urlMatchPattern.pattern === MATCH_ALL_URLS;
+        });
```

Comment:
The reason I chose to do this over `.test` is that, it is cheaper to do an equality check on all items. 

Like you said it will have an overhead if the list in hundreds without a *MATCH_ALL_URLS*. But we don't do complex operation for the test. It should be acceptable.

Imagine a long list with *MATCH_ALL_URLS* at the end, then we would have to do slow regex test on all items till we get there.

However, I am not totally poised to this approach, because what is the chance of having a long list with *MATCH_ALL_URLS* at the end. So not sure if we need to optimise for this path.